generator/statistics.cpp

```diff
@@ -53,22 +55,55 @@ namespace stats
 
     void operator() (FeatureType & f, uint32_t)
     {
-      f.ParseBeforeStatistic();
+      f.ParseHeader2();
 
-      FeatureType::InnerGeomStat const innerStats = f.GetInnerStatistic();
+      FeatureType::InnerGeomStat const innerStats = f.GetInnerStats();
 
       m_info.m_inner[0].Add(innerStats.m_points);
       m_info.m_inner[1].Add(innerStats.m_strips);
       m_info.m_inner[2].Add(innerStats.m_size);
 
-      // get geometry size for the best geometry
-      FeatureType::GeomStat const geom = f.GetGeometrySize(FeatureType::BEST_GEOMETRY);
-      FeatureType::GeomStat const trg = f.GetTrianglesSize(FeatureType::BEST_GEOMETRY);
+      // Get size stats and load the best geometry.
+      FeatureType::GeomStat const geom = f.GetOuterGeometryStats();
+      FeatureType::GeomStat const trg = f.GetOuterTrianglesStats();
 
-      m_info.m_byPointsCount[CountType(geom.m_count)].Add(geom.m_size);
-      m_info.m_byTrgCount[CountType(trg.m_count / 3)].Add(trg.m_size);
+      uint32_t outerGeomSize = 0, outerTrgSize = 0;
+      for (size_t ind = 0; ind < DataHeader::kMaxScalesCount; ++ind)
+      {
+        auto const geomSize = geom.m_sizes[ind], geomElems = geom.m_elements[ind],
+                   trgSize = trg.m_sizes[ind], trgElems = trg.m_elements[ind];
+        m_info.m_byLineGeom[ind].Add(geomSize, geomElems);
+        outerGeomSize += geomSize;
+        m_info.m_byAreaGeom[ind].Add(trgSize, trgElems);
+        outerTrgSize += trgSize;
+
+        if (ind > 0)
+        {
+          // If a feature has a more simplified version of current geometry.
+          if (geom.m_elements[ind - 1] > 0)
+            m_info.m_byLineGeomCompared[ind].Add(geomSize, geomElems);
+          if (trg.m_elements[ind - 1] > 0)
+            m_info.m_byAreaGeomCompared[ind].Add(trgSize, trgElems);
+        }
+
+        if (ind < DataHeader::kMaxScalesCount - 1)
+        {
+          // If feature's current geometry almost duplicates a more detailed one
+          // (has <geometryDupFactor less elements).
+          if (geomElems > 0 && geomElems * m_info.m_geometryDupFactor > geom.m_elements[ind + 1])
+            m_info.m_byLineGeomDup[ind].Add(geomSize, geomElems);
+          if (trgElems > 0 && trgElems * m_info.m_geometryDupFactor > trg.m_elements[ind + 1])
+            m_info.m_byAreaGeomDup[ind].Add(trgSize, trgElems);
+        }
+      }
```

Comment:
Can't realize,  how m_byLineGeomCompared and m_byLineGeomDup will help us to understand something? Ok, we have 5M bytes of geometry that have previous simplified geometry, and?

Why not to collect and print histogram in range of [0.1, 0.2, ..., 0.9, 1.0]?
where factor = geom. m_sizes[ind - 1] / geom. m_sizes[ind] ?

Reply:
```
trg2: size =   2075253: elements =    545051; features =   18169; elems/feats =  530.0; bytes/elems =  3.8
trg3w/2: size =   4096561: elements =   1174955; feats w/trg2 =   18169; elems/feats =  64.7; size factor =  2.0x; elems factor =  2.2x
   trg3: size =   4299267: elements =   1227069; features =   24519; elems/feats =  50.0; bytes/elems =  3.5
```
`m_byLineGeomCompared` shows us that 4.1Mb or 1.2M of triangles were simplified into 2.1Mb / 0.5M. So overall simplification effectiveness was ~2x (which is not much compared to trg1/trg0 difference).

The `factor = geom. m_sizes[ind - 1] / geom. m_sizes[ind]` is misleading, because it includes e.g. features with trg3 without trg2, i.e. they were not simplified. Hence we can't judge simplification effectiveness looking at this factor alone.

`m_byLineGeomDup` shows percentage of very similar geometries. It gives an idea of potential size savings if we decide to omit such similar geometries.

A histogram is a great idea and will give much more info that one number, but its just takes more time to implement :)